src/xml_node.cc

```diff
@@ -230,66 +230,257 @@ XmlNode::New(xmlNode* node)
+                }
+            }
+
+        }
+
+        if (wrapped_node != NULL) {
+            break;
+        }
+
+        xml_obj = next;
+    }
+
+    return wrapped_node;
+}
+
+/*
+ * Search descendants of node to find javascript wrapper,
+ * optionally ignoring given node. Based on xmlFreeNode.
+ */
+xmlNode*
+get_wrapped_descendant(xmlNode *xml_obj, xmlNode *skip_xml_obj=NULL) {
+
+    xmlNode* wrapped_descendant = NULL;
+
+    if (xml_obj->type == XML_DTD_NODE) {
+        return (xml_obj->children == NULL) ?
+            NULL : get_wrapped_node_in_children(xml_obj->children, skip_xml_obj);
+    }
+
+    if (xml_obj->type == XML_NAMESPACE_DECL) {
+        return NULL;
+    }
+
+    if (xml_obj->type == XML_ATTRIBUTE_NODE) {
+        return (xml_obj->children == NULL) ?
+            NULL : get_wrapped_node_in_children(xml_obj->children, skip_xml_obj);
+    }
+
+    if ((xml_obj->children != NULL) && (xml_obj->type != XML_ENTITY_REF_NODE)) {
+        wrapped_descendant =
+            get_wrapped_node_in_children(xml_obj->children, skip_xml_obj);
+    }
+
+    if ((xml_obj->type == XML_ELEMENT_NODE) ||
+        (xml_obj->type == XML_XINCLUDE_START) ||
+        (xml_obj->type == XML_XINCLUDE_END)) {
+
+        if ((wrapped_descendant == NULL) && (xml_obj->properties != NULL)) {
+            wrapped_descendant =
+                reinterpret_cast<xmlNode*>(get_wrapped_attr_in_list(xml_obj->properties, skip_xml_obj));
+        }
+
+        if ((wrapped_descendant == NULL) && (xml_obj->nsDef != NULL)) {
+            wrapped_descendant =
+                reinterpret_cast<xmlNode*>(get_wrapped_ns_in_list(xml_obj->nsDef, skip_xml_obj));
+        }
+    }
+
+    return wrapped_descendant;
+}
+
+XmlNode::~XmlNode() {
+    if ((this->doc != NULL) && (this->doc->_private != NULL)) {
+      static_cast<XmlDocument*>(this->doc->_private)->Unref();
+    }
+    this->unref_wrapped_ancestor();
+    if (xml_obj == NULL) return;
 
     xml_obj->_private = NULL;
-    // release the hold and allow the document to be freed
-    XmlDocument* doc = static_cast<XmlDocument*>(xml_obj->doc->_private);
-    doc->unref();
-
-    // We do not free the xmlNode here if it is linked to a document
-    // It will be freed when the doc is freed
-    if (xml_obj->parent == NULL)
-      xmlFreeNode(xml_obj);
-
-    // if there's a parent then Unref() it
-    else if (xml_obj->parent->_private != NULL &&
-            (void*)xml_obj->doc != (void*)xml_obj->parent)
-    {
-        XmlNode* parent = static_cast<XmlNode*>(xml_obj->parent->_private);
-
-        // make sure Unref() is necessary
-        if (parent->refs_ > 0)
-        {
-            parent->Unref();
+    if (xml_obj->parent == NULL) {
+        if (get_wrapped_descendant(xml_obj) == NULL) {
+            xmlFreeNode(xml_obj);
         }
     }
+    else {
+        xmlNode *ancestor = get_wrapped_ancestor_or_root(xml_obj);
```

Comment:
It might be possible to avoid this lookup if we already know that `this->ancestor` was not null. We could make `this->unref_wrapped_ancestor()` return true if it unrefed an ancestor.

So we could then say something like `hasWrappedAncestor = this->unref_wrapped_ancestor()`, then make the condition `} else if (!hasWrappedAncestor) { ...`. This way we avoid another lookup if we already know about a wrapped ancestor.